numba/np/linalg.py

```diff
@@ -521,39 +520,84 @@ def check_args(a, b):
     return builder.load(out)
 
 
-@glue_lowering(np.dot, types.Array, types.Array)
-def dot_2(context, builder, sig, args):
+@overload(np.dot)
+def dot_2(left, right):
     """
     np.dot(a, b)
+    """
+    return dot_2_impl('np.dot()', left, right)
+
+
+@overload(operator.matmul)
+def matmul_2(left, right):
+    """
     a @ b
     """
-    ensure_blas()
-
-    with make_contiguous(context, builder, sig, args) as (sig, args):
-        ndims = [x.ndim for x in sig.args[:2]]
-        if ndims == [2, 2]:
-            return dot_2_mm(context, builder, sig, args)
-        elif ndims == [2, 1]:
-            return dot_2_mv(context, builder, sig, args)
-        elif ndims == [1, 2]:
-            return dot_2_vm(context, builder, sig, args)
-        elif ndims == [1, 1]:
-            return dot_2_vv(context, builder, sig, args)
-        else:
-            assert 0
+    return dot_2_impl("'@'", left, right)
+
+
+def dot_2_impl(name, left, right):
+    if isinstance(left, types.Array) and isinstance(right, types.Array):
+        @intrinsic
+        def _impl(typingcontext, left, right):
+            ndims = (left.ndim, right.ndim)
+
+            def _dot2_codegen(context, builder, sig, args):
+                ensure_blas()
 
+                with make_contiguous(context, builder, sig, args) as (sig, args):
+                    if ndims == (2, 2):
+                        return dot_2_mm(context, builder, sig, args)
+                    elif ndims == (2, 1):
+                        return dot_2_mv(context, builder, sig, args)
+                    elif ndims == (1, 2):
+                        return dot_2_vm(context, builder, sig, args)
+                    elif ndims == (1, 1):
+                        return dot_2_vv(context, builder, sig, args)
+                    else:
+                        assert 0
```

Comment:
```suggestion
                        raise AssertionError('unreachable')
```
because Python can skip `assert` statement with `python -O <script>`

Reply:
No problem. There's actually a bunch of these assertions in this file.  This one was in the original code block. Maybe some of the others need to be replaced.